src/NHibernate/Linq/DmlExpressionRewriter.cs

```diff
@@ -38,6 +38,35 @@ void AddSettersFromBindings(IEnumerable<MemberBinding> bindings, string path)
 				}
 		}
 
+		void AddSettersFromConstructorArguments(NewExpression newExpression, string path)
+		{
+			// See Members documentation, this property is specifically designed to match constructor arguments values
+			// in the anonymous object case. It can be null otherwise, or non-matching.
+			var argumentMatchingMembers = newExpression.Members;
+			if (argumentMatchingMembers == null || argumentMatchingMembers.Count != newExpression.Arguments.Count)
```

Comment:
There are constructor paths setting it to null (seen with IlSpy). About non-matching, I have not seen an explicit case, but nothing seems to be there for preventing building manually a `NewExpression` with a non-matching members list. So I have put at least a check for not failing with an index out of range.